### PR TITLE
Default Modifier

### PR DESCRIPTION
**Pull request by SergioRibera**
Create a new default modifier param and make keybind modifier optional, based on #9 
Create test for the new parser of config

- [x] feature
- [x]  tests

> Note: after talking about this feature, I think this feature is necessary for future features.

### CODE REVIEW

File: lefthk/src/config.rs [hunk 1]
```diff
@@ -175,3 +181,140 @@ fn propagate_exit_chord(chords: Vec<&mut Keybind>, exit_chord: Option<Keybind>)
...
+        assert_eq!(
+            custom_keybind.command,
+            Command::Chord(vec![Keybind {
+                command: Command::Execute("st -e htop".to_string()),
+                modifier: vec!["Mod4".to_string()],
+                key: "c".to_string(),
+            }])
+        );
+    }
+}
```
Comment (AethanFoot) [inline lefthk/src/config.rs:320]: Could you move this test to test.rs

Comment (SergioRibera) [reply lefthk/src/config.rs]: Sorry for the delay, I just uploaded an update where I moved the tests as you requested

**Comment by AethanFoot** [comment]
Ill give this a test asap

File: lefthk/src/config.rs [hunk 2]
```diff
@@ -42,101 +42,100 @@ macro_rules! get_keys {
...
 }
 
-pub fn load() -> Result<Config> {
-    let path = BaseDirectories::with_prefix(lefthk_core::LEFTHK_DIR_NAME)?;
-    fs::create_dir_all(&path.get_config_home())?;
-    let file_name = path.place_config_file("config.ron")?;
-    if Path::new(&file_name).exists() {
-        let contents = fs::read_to_string(file_name)?;
+impl Config {
+    pub fn from_string(contents: String) -> Result<Self> {
```
Comment (TornaxO7) [inline lefthk/src/config.rs:137]: wouldn't it be better to make use of the `TryFrom` trait?

Comment (TornaxO7) [inline lefthk/src/config.rs:137]: wouldn't it be better to make use of the `TryFrom` trait?

File: lefthk/src/config.rs [hunk 3]
```diff
@@ -42,101 +42,100 @@ macro_rules! get_keys {
...
-                    .filter_map(|kb| match TryFrom::try_from(kb.clone()) {
-                        Ok(keybinds) => Some::<Vec<lefthk_core::config::Keybind>>(keybinds),
-                        Err(err) => {
-                            tracing::error!("Invalid key binding: {}\n{:?}", err, kb);
-                            None
-                        }
-                    })
-                    .flatten()
-                    .collect();
+fn try_from(kb: Keybind, default_modifier: Vec<String>) -> Result<Vec<core_keybind>> {
```
Comment (TornaxO7) [inline lefthk/src/config.rs:49]: I'd rename this to `try_convert_to_core_keybinds` or something like that

**Review by TornaxO7 — CHANGES_REQUESTED** [review]
Added some suggestions

File: lefthk/src/config.rs (hunk 2 above)
Comment (SergioRibera) [reply lefthk/src/config.rs]: This suggestion pleases me, thank you very much, I have already solved it

Comment (TornaxO7) [reply lefthk/src/config.rs]: :+1:

File: lefthk/src/config.rs (hunk 3 above)
Comment (SergioRibera) [reply lefthk/src/config.rs:49]: This one here, I like it but I don't want to give it a very long name

Comment (SergioRibera) [reply lefthk/src/config.rs:49]: @TornaxO7 what do you think about `try_to_core_keybinds`?

Comment (TornaxO7) [reply lefthk/src/config.rs:49]: understandable. Hm.... In my opinion this big function needs a refactoring anyhow. But this can be done in another PR.

**Review by TornaxO7 — APPROVED** [review]
LGTM

File: lefthk/src/config.rs (hunk 3 above)
Comment (SergioRibera) [reply lefthk/src/config.rs:49]: I think the same, if you need help, I am available to collaborate

Comment (TornaxO7) [reply lefthk/src/config.rs:49]: > @TornaxO7 what do you think about `try_to_core_keybinds`?

yah! That sonunds better

Comment (TornaxO7) [reply lefthk/src/config.rs:49]: @SergioRibera I started to create a [PR](https://github.com/leftwm/lefthk/pull/29) for this.

Comment (SergioRibera) [reply lefthk/src/config.rs:49]: Great, let me see how that PR goes and see what I can do to help

**Comment by VuiMuich** [comment]
Not enough time to fix the broken test rn, maybe later tonight.

**Comment by SergioRibera** [comment]
@VuiMuich No problem, I'll check right away what is wrong with the CI processes and do some push to fix it

**Comment by VuiMuich** [comment]
> @VuiMuich No problem, I'll check right away what is wrong with the CI processes and do some push to fix it

Thanks for returning to this PR so promptly after that long time 😃

Nice idea to generate the normalized command from the raw string.

**Comment by SergioRibera** [comment]
> Thanks for returning to this PR so promptly after that long time 😃

Of course, now I try to be more attentive to all this, I will be reviewing the project to see if I can find more I can collaborate.

> If you have a minute to `cargo fmt` that would be nice, otherwise I could also do this before merging.

Oh no, I could have sworn I ran the corresponding commands before pushing, I'll fix it right away, thank you very much

**Comment by VuiMuich** [comment]
Thanks for your contribution. If you have more to come this is surely welcome. Hopefully we don't have you waiting for the merge for so long again.

**Pull request merged**